python/pyarrow/src/arrow/python/udf.cc

```diff
@@ -105,21 +158,117 @@ Status RegisterScalarFunction(PyObject* user_function, ScalarUdfWrapperCallback
+      wrapper, options, registry);
+}
 
+namespace  {
+
+Result<std::shared_ptr<RecordBatch>> RecordBatchFromArray(
+    std::shared_ptr<Schema> schema, std::shared_ptr<Array> array) {
+  auto& data = const_cast<std::shared_ptr<ArrayData>&>(array->data());
+  if (data->child_data.size() != static_cast<size_t>(schema->num_fields())) {
+    return Status::Invalid("UDF result with shape not conforming to schema");
+  }
+  return RecordBatch::Make(std::move(schema), data->length, std::move(data->child_data));
+}
+
+}  // namespace
+
+Result<std::shared_ptr<RecordBatchReader>> CallTabularFunction(
+    const std::string& func_name, const std::vector<Datum>& args,
+    compute::FunctionRegistry* registry) {
+  if (args.size() != 0) {
+    return Status::NotImplemented("non-empty arguments to tabular function");
+  }
+  if (registry == NULLPTR) {
+    registry = compute::GetFunctionRegistry();
+  }
+  ARROW_ASSIGN_OR_RAISE(auto func, registry->GetFunction(func_name));
+  if (func->kind() != compute::Function::SCALAR) {
+    return Status::Invalid("tabular function of non-scalar kind");
+  }
+  auto arity = func->arity();
+  if (arity.num_args != 0 || arity.is_varargs) {
+    return Status::NotImplemented("tabular function of non-null arity");
+  }
+  auto kernels =
+      arrow::internal::checked_pointer_cast<compute::ScalarFunction>(func)->kernels();
+  if (kernels.size() != 1) {
+    return Status::NotImplemented("tabular function with non-single kernel");
+  }
+  const compute::ScalarKernel* kernel = kernels[0];
+  auto out_type = kernel->signature->out_type();
+  if (out_type.kind() != compute::OutputType::FIXED) {
+    return Status::Invalid("tabular kernel of non-fixed kind");
+  }
+  auto datatype = out_type.type();
+  if (datatype->id() != Type::type::STRUCT) {
+    return Status::Invalid("tabular kernel with non-struct output");
+  }
+  auto struct_type = arrow::internal::checked_cast<StructType*>(datatype.get());
+  auto schema = ::arrow::schema(struct_type->fields());
+  std::vector<TypeHolder> in_types;
+  ARROW_ASSIGN_OR_RAISE(auto func_exec,
+                        GetFunctionExecutor(func_name, in_types, NULLPTR, registry));
+  auto next_func =
+      [schema,
+       func_exec = std::move(func_exec)]() -> Result<std::shared_ptr<RecordBatch>> {
+    std::vector<Datum> args;
+    // passed_length of -1 or 0 with args.size() of 0 leads to an empty ExecSpanIterator
+    // in exec.cc and to never invoking the source function, so 1 is passed instead
+    ARROW_ASSIGN_OR_RAISE(auto datum, func_exec->Execute(args, /*passed_length=*/1));
+    if (!datum.is_array()) {
+      return Status::Invalid("UDF result of non-array kind");
+    }
+    std::shared_ptr<Array> array = datum.make_array();
+    if (array->length() == 0) {
+      return IterationTraits<std::shared_ptr<RecordBatch>>::End();
+    }
+    return RecordBatchFromArray(std::move(schema), std::move(array));
```

Comment:
Right now it seems a tabular function must return a struct array.  I think that is "ok" but for consistency with the rest of the kernels might it be better to return an ExecSpan?

Of course, we have no concept of exec span in python so the python UDF would still return a struct array.  The C++ wrapper could then convert that returned struct array into an exec span.

Although, continuing down this line, would it make more sense to have the python UDFs return record batches?  pyarrow doesn't really have much concept / functionality for struct array.

Reply:
I'll look into this. I vaguely remember that I had some reason for using `StructArray`, but maybe this can be worked out.

Reply:
Reviewing this, I believe I used `StructArray` because the [original result-handling code](https://github.com/apache/arrow/blob/5c1044fce55ed0e373a622cb8ee3b97a1a34799a/python/pyarrow/src/arrow/python/udf.cc#L68-L83) deals with arrays only, so introducing a new type of result should probably be done with care and consistency. You should know this original code - WDYT?

Reply:
Right, I think we are balancing between:

 * Array-in / Array-out is a simple model so tabular functions can return arrays
 * Struct arrays aren't used much in pyarrow but record batches are

I think there are merits to both.  However, I don't feel too strongly about either choice.  So we can always leave it how it is and adjust if needed.  This feature will still be fairly experimental for a while I think.